File: Easy/valid_parentheses/valid_parentheses.py
from typing import List

def isValid(s: str) -> bool:
    stack = []
    for char in s:
        if char in "([{":
            stack.append(char)  
        else:
            # If stack is empty or the top of the stack doesn't match the corresponding opening bracket, return False
            if not stack:
                return False
            top = stack.pop()
            if char == ')' and top != '(':
                return False
            if char == ']' and top != '[':
                return False
            if char == '}' and top != '{':
                return False
            
    return len(stack) == 0  # If stack is empty, all brackets are valid

if __name__ == "__main__":
    print("Testing isValid function:")
    print("For the input '()', the output is:", isValid("()")) 
    print("For the input '()[]{}', the output is:", isValid("()[]{}"))
    print("For the input '(]', the output is:", isValid("(]"))
    print("For the input '([])', the output is:", isValid("([])"))
    print("For the input '([)]', the output is:", isValid("([)]"))

File: Easy/valid_parentheses/valid_parentheses.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stack>
using namespace std;

class Solution {
    public:
        bool isValid(string s) {
            stack<char> bracketStack;
            for(int i = 0; i < s.size(); i++) {
                char c = s[i];
                if(c == '(' || c == '{' || c == '[') {
                    bracketStack.push(c);
                }
                else {
                    if(bracketStack.empty()) {
                        return false;
                    }
                    char top = bracketStack.top();
                    bracketStack.pop();
                    if((c == ')' && top != '(') || (c == '}' && top != '{') || (c == ']' && top != '[')) {
                        return false;
                    }
                }
            }
            return bracketStack.empty();
        }
};

//Example usage:
int main() {
    cout << boolalpha; //Print boolean values as true/false
    cout << "Testing Valid Parentheses: " << endl;
    Solution solution1;
    string s1 = "()";
    cout << "Input: " << s1 << " - Output: " << solution1.isValid(s1) << endl;

    string s2 = "()[]{}";
    cout << "Input: " << s2 << " - Output: " << solution1.isValid(s2) << endl;

    string s3 = "(]";
    cout << "Input: " << s3 << " - Output: " << solution1.isValid(s3) << endl;

    string s4 = "([)]";
    cout << "Input: " << s4 << " - Output: " << solution1.isValid(s4) << endl;

    string s5 = "{[]}";
    cout << "Input: " << s5 << " - Output: " << solution1.isValid(s5) << endl;  
}